include/tc/core/execution_engine-inl.h

```diff
@@ -0,0 +1,228 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "tc/core/polyhedral/mapping_types.h"
+#include "tc/core/utils/memory.h"
+
+#include "tc/lang/parser.h"
+
+namespace tc {
+
+// Under object lock, fill parse the language and fill the underlying map
+template <typename ExecutorType>
+void ExecutionEngine<ExecutorType>::define(const std::string& language) {
+  lang::Parser parser(language);
+  std::lock_guard<std::mutex> lg(tcExecutorMutex_);
+  while (parser.L.cur().kind != lang::TK_EOF) {
+    auto treeRef = parser.parseFunction();
+    auto name = lang::Def(treeRef).name().name();
+    tcNameMap_.emplace(std::make_pair(name, treeRef));
+  }
+}
+
+// support define if we pass the parsed TreeRefs.
+template <typename ExecutorType>
+void ExecutionEngine<ExecutorType>::define(
+    const std::vector<lang::TreeRef>& treeRefs) {
+  std::lock_guard<std::mutex> lg(tcExecutorMutex_);
+  for (auto& ref : treeRefs) {
+    auto name = lang::Def(ref).name().name();
+    tcNameMap_.emplace(std::make_pair(name, ref));
```

Comment:
let's try to reuse the code.

Reply:
@nicolasvasilache do you plan to reuse the code? if not, can you explain why is it better to keep same code in two places which are doing essentially same thing :)

Reply:
yup getting to those in order of commits, I was hoping to finish yesterday before leaving but got interrupted ;)

Reply:
addressed in c7effdb